Fix: change base class of `SimpleArrayData` from `_XMLObject` to `_BaseObject`

hirohira9119 opened the pull request:
### **User description**
This change reflects the fact that `gx:SimpleArrayData` extends `kml:AbstractObjectType` in the XSD definition, which corresponds to `_BaseObject` in the codebase.
Also added missing `kwargs` to the constructor.
 Sorry I missed that in the previous pull request!


___

### **PR Type**
Bug fix


___

### **Description**
- Change `SimpleArrayData` base class from `_XMLObject` to `_BaseObject`

- Add missing `id` and `target_id` parameters to constructor

- Update constructor to accept `**kwargs` parameter

- Enhance docstrings and method documentation


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["_XMLObject"] -- "inheritance fix" --> B["_BaseObject"]
  C["SimpleArrayData"] -- "inherits from" --> B
  D["Constructor"] -- "add parameters" --> E["id, target_id, **kwargs"]
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>data.py</strong><dd><code>Fix inheritance and constructor parameters</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

fastkml/gx/data.py

<ul><li>Changed <code>SimpleArrayData</code> base class from <code>_XMLObject</code> to <code>_BaseObject</code><br> <li> Added <code>id</code> and <code>target_id</code> parameters to constructor<br> <li> Added <code>**kwargs</code> parameter handling<br> <li> Enhanced docstrings with proper formatting and type hints</ul>


</details>


  </td>
  <td><a href="https://github.com/cleder/fastkml/pull/445/files#diff-fbc57e29efecedc93ac51f70eb22f973ea5f891150c9a8fef5ca776b029c3c45">+44/-9</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>data_test.py</strong><dd><code>Update tests for new constructor parameters</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/hypothesis/gx/data_test.py

<ul><li>Added <code>id</code> and <code>target_id</code> parameters to test function<br> <li> Updated test to include new constructor parameters<br> <li> Added import for <code>nc_name</code> strategy</ul>


</details>


  </td>
  <td><a href="https://github.com/cleder/fastkml/pull/445/files#diff-2afd30eceac98e07261aa7b22528845a3c5140315594af6148f1e7c7db595a91">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___


<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Change `SimpleArrayData` base class to `_BaseObject` and update constructor and tests for consistency with XSD definition.
> 
>   - **Class Changes**:
>     - Change base class of `SimpleArrayData` from `_XMLObject` to `_BaseObject` in `data.py`.
>     - Add `kwargs` to `SimpleArrayData` constructor for flexible initialization.
>   - **Testing**:
>     - Update `test_fuzz_simple_array_data` in `data_test.py` to include `id` and `target_id` parameters.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cleder%2Ffastkml&utm_source=github&utm_medium=referral)<sup> for d1a392c4ddde5040144506ca8de3d3ff53ddebc3. You can [customize](https://app.ellipsis.dev/cleder/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

## Summary by Sourcery

Fix inheritance and constructor signature of SimpleArrayData to match the XSD definition and update tests accordingly

Bug Fixes:
- Align SimpleArrayData inheritance by changing its base class to _BaseObject
- Add missing id, target_id, and **kwargs parameters to SimpleArrayData constructor

Enhancements:
- Enhance docstrings for SimpleArrayData __init__, __repr__, and __bool__ methods

Tests:
- Update fuzz tests to include id and target_id parameters and import nc_name strategy

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - SimpleArrayData now accepts optional id and target_id fields.
  - String representation includes id and target_id for clearer debugging.

- Documentation
  - Expanded docstrings to cover new fields and clarify boolean behavior.

- Tests
  - Property-based tests updated to generate and validate id and target_id inputs.
  - Test environment extended to support tzwin-based repr evaluations.

- Refactor
  - SimpleArrayData aligned to a new internal base type; import path updated for consumers.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hirohira9119 committed:
Fix: change base class of SimpleArrayData from _XMLObject to _BaseObject

cleder commented:
mypy reports some errors it would be great if they could be resolved
https://github.com/cleder/fastkml/actions/runs/17585583758/job/49952562440?pr=445
Great work :-)

cleder commented:
I have to find some time to resolve the hypothesis errors :-(

hirohira9119 committed:
Fix mypy error

hirohira9119 committed:
Fix another mypy error

hirohira9119 committed:
Fix NameError in hypothesis/gx/track_test for windows

hirohira9119 commented:
Thank you for the review.

Apologies for the oversight—there were no errors reported by Pylance in my local environment, which led me to miss the issue. I have now addressed the mypy error as suggested.

In addition, although these changes are outside the scope of the previous commit, I have included two minor fixes in this PR. If you would prefer these to be submitted separately, I am happy to open a new PR accordingly.

- A mypy error had been present in `container.py` since a recent revision, likely triggered by the upgrade from mypy 1.16.1 to 1.17.1. I have adjusted the placement of `# type: ignore` to resolve the issue.
- In `hypothesis/gx/track_test.py`, a `NameError` could occur on Windows due to the use of the `tzwin` class without an explicit import. I have added the necessary import for `dateutil.tz.tzwin`. Since `tzwin` is defined as `None` on non-Windows platforms, this fix has been applied universally.

hirohira9119 commented:
I'd like to share some findings from my own investigation into the Hypothesis-related errors. Please feel free to disregard this if you're already aware of these issues.

I've identified two separate cases:

---

(1) Issue in `assert_str_roundtrip_terse`

This error occurs in several classes due to the following mechanism. I'll use `Region` as an example:

1. In some cases, the `Region` object includes a parameter like `Lod(min_lod_pixels=256, max_lod_pixels=0)`. When serialized with `to_string(terse)`, this becomes `<Lod><max_lod_pixels>0</max_lod_pixels></Lod>`, since `min_lod_pixels` matches its default value and is skipped in `terse` mode.
2. When deserialized via `from_string()`, the resulting object is `Lod(max_lod_pixels=0).`
3. When re-serialized with `to_string(verbose)`, the entire `Lod` element is skipped. This is because `Lod` evaluates to `False` when `min_lod_pixels is None`, and thus is not appended by `xml_subelement()`.
4. As a result, the `verbose` output differs from the original, causing the test to fail.

In short, this issue arises when a subelement is an `XMLObject` that evaluates to falsy. As a potential fix, I suggest modifying step 3 so that in `verbose` mode, any non-`None` value is included in the output.

Specifically, the following change to `xml_subelement` (and similarly to `xml_subelement_list`) resolves the issue:

```python
value = getattr(obj, attr_name, None)
if value or (verbosity == Verbosity.verbose and value is not None):
    element.append(
        value.etree.etree_element(
            precision=precision,
            verbosity=verbosity,
        ),
    )
```

I’ve confirmed that this resolves the class of errors described above. If you think this approach is appropriate, I’d be happy to open a PR—just let me know.

---

(2) Issue in `KmlDateTime`

In `assert_str_roundtrip`, object comparison may fail due to loss of timezone information when passing through string conversion.

```python
kdt = KmlDateTime(dt=datetime.datetime(2016, 10, 30, 3, 0, tzinfo=dateutil.tz.gettz('Europe/Helsinki')))
print(repr(kdt))  # has timezone info
# fastkml.times.KmlDateTime(dt=datetime.datetime(2016, 10, 30, 3, 0, tzinfo=tzfile('Europe/Helsinki')), resolution=DateTimeResolution.datetime, )
print(str(kdt))   # no timezone info
# 2016-10-30T03:00:00+03:00
print(KmlDateTime.parse(str(kdt)))  # no timezone info
# fastkml.times.KmlDateTime(dt=datetime.datetime(2016, 10, 30, 3, 0, tzinfo=tzoffset(None, 10800)), resolution=DateTimeResolution.datetime, )
```

I don’t currently have a solution for this, but I wanted to share it as a known issue that I’ve encountered.

[Edit note]
Considered two approaches to address the `KmlDateTime` timezone normalization issue:
A. Normalize `datetime` in the `KmlDateTime` constructor by replacing `tzinfo` with `timezone(utcoffset)` when dt is `datetime` and `tzinfo` is not `None`.
B. In `KmlDateTime.__eq__`, explicitly compare datetime components (`year`, `month`, ..., `microsecond`) and `utcoffset()` when comparing `datetime` instances.

Either approach resolves the issue. Option A offers simpler code, while option B may provide slightly better performance. (But the difference is minor).

Sample code for (1), (2)-A, and (2)-B has been published as shown below.  
https://github.com/hirohira9119/fastkml/compare/e3ec94d...eb5e57e

All tests including Hypothesis currently pass 100%.  
If needed, the implementation can be refined further and submitted as a PR.

Note: Either (2)-A or (2)-B alone is sufficient to pass the tests, but both have been committed for illustrative purposes.

cleder commented:
I like the B approach to date times `__eq__` 
after all if it is the same datetime it should be considdered equal. 
Another approach might be to normalize the date-times to UTC  (in `__eq__`) and compare if they are the same point in time.
What do you think?

cleder commented:
>  this issue arises when a subelement is an XMLObject that evaluates to falsy

I'd prefer to adjust the test generation data for the `Lod` so it can never be falsy and test this special case as a normal unit test.

Background: An XMLElement evaluates to false if the information in this element does not provide any information other than the default values, so it does not matter to a client if it is present or not.

hirohira9119 commented:
Thank you for merging :)

I appreciate your feedback on the `KmlDateTime` issue. I agree that approach B is cleaner, as it preserves the original `dt`.

Regarding the comparison logic, I now believe that normalizing to UTC is the appropriate solution. According to the XML Schema specification, equality should be determined based on the actual point in time. Comparing in UTC aligns with this definition and improves both correctness and code clarity.

I’ve opened https://github.com/cleder/fastkml/pull/446 to reflect this—happy to continue the discussion there.

One note: I additionaly found that the issue only occurs during the DST fall back period. Outside that, `datetime.__eq__` behaves as expected. This likely explains the low failure rate in Hypothesis tests.

Since handling DST separately adds complexity and UTC conversion is lightweight, I’ve opted to apply it consistently.

hirohira9119 commented:
> I'd prefer to adjust the test generation data for the `Lod` so it can never be falsy and test this special case as a normal unit test.
> 
> Background: An XMLElement evaluates to false if the information in this element does not provide any information other than the default values, so it does not matter to a client if it is present or not.

Thank you for the explanation. It's clear now. After applying a filter to the relevant parameters instead of changing `xml_subelement()`, all Hypothesis tests are now passing consistently. I've opened a draft PR based on that: https://github.com/cleder/fastkml/pull/447.
Adding the special cases to normal unit tests will follow.
I'd suggest continuing the discussion in that PR thread.
